### ISSUE TITLE
Separate log and container lifecycle management

### ISSUE THREAD

**Issue by HirazawaUi** [kubernetes/kubernetes#127376]
This is a follow-up to issue #123279, as the scope of the discussion has expanded beyond the original issue. Therefore, I am opening a new issue to focus on this topic.

In issue #123279, the goal was to disable kubelet from removing container logs when the container is deleted. However, disabling kubelet from cleaning up logs of removed containers could introduce potential risks, such as log files consuming excessive disk space, eventually leading to the host running out of storage and becoming unresponsive.

With that in mind, I would like to further discuss a topic we had previously considered: **separating the lifecycle management of containers and their logs in kubelet**, with the retention of logs for deleted containers being a subset of this broader topic.
ref: https://github.com/kubernetes/kubernetes/blob/af2bf2d8f1b3437fe67e1686266a176320422465/pkg/kubelet/kuberuntime/kuberuntime_container.go#L1308-L1314

Below are some features that I think:

1. **Rotation and Compression of Logs from Deleted Pods**: Logs for containers are stored by default under /var/log/pods/, with a unique directory structure based on the pod's namespace-podName-podUUID. Perhaps we should consider compressing the entire pod log directory when the pod is deleted?

2. **Kubelet Management of Deleted Pod Logs**: Kubelet should have a clear configuration to manage the cleanup and retention policy for logs from deleted pods. We could introduce new settings, such as **podLogMaxFiles** and **podLogMaxSize**, similar to the existing containerLogMaxFiles and containerLogMaxSize.

3. **Capacity Limits for Pod Log Directories**: Should we enforce a maximum storage limit for pod log directories? If no limit is set, they could potentially fill up the entire host storage.

I'm not entirely sure if these ideas are on the right track, or if simply disabling kubelet's log cleanup for containers is sufficient. I would greatly appreciate any feedback or suggestions.

/sig node
/kind feature

**Comment by CpuID** [comment]
re https://github.com/kubernetes/kubernetes/issues/123279 - I don't know how safe it would be to disable log cleanup entirely (for the runaway disk space reasons already discussed)

I feel like the safer answer would be if kubelet still has to be involved in log cleanup, but has the ability to delay cleanup/deletion for a configurable (but with a hardcoded upper bound) amount of time, say up to 90~ seconds?

There would likely need to be fail-safes added to that, eg. if a node's disk space were to fill in that delay window, prioritise node reliability and GC the logs before the desired cleanup time, maybe name it as a "maximum log cleanup delay" with associated docs, to avoid any unintended assumptions by users.

I raised https://github.com/kubernetes/kubernetes/issues/130830 to discuss "allow log files to stick around after a pod is deleted for a configurable amount of time" but this is relevant the wider log lifecycle management piece, happy to move that conversation here if it doesn't pollute/distract things too much :)

-----------------

some feedback re the other features flagged for discussion:

> Rotation and Compression of Logs from Deleted Pods: Logs for containers are stored by default under /var/log/pods/, with a unique directory structure based on the pod's namespace-podName-podUUID. Perhaps we should consider compressing the entire pod log directory when the pod is deleted?

If implemented, this should be configurable (boolean on/off).

I don't think this would be super beneficial for "external log shipper" purposes, unless you're talking a substantially large size of a log in a short duration (theoretically possible). from a log shippers perspective, that file becoming compressed is another opportunity for that shipper to need to rescan for available log files. being most log shippers tend to do this rescan between 1-60 second intervals, the benefits may not be there...

> Kubelet Management of Deleted Pod Logs: Kubelet should have a clear configuration to manage the cleanup and retention policy for logs from deleted pods. We could introduce new settings, such as podLogMaxFiles and podLogMaxSize, similar to the existing containerLogMaxFiles and containerLogMaxSize.

I do think this is valid, and goes with my suggestion of a maximum amount of time to wait before cleanup 👍 max files/size/delay time are all valid configurable options to consider.

> Capacity Limits for Pod Log Directories: Should we enforce a maximum storage limit for pod log directories? If no limit is set, they could potentially fill up the entire host storage.

probably a valid consideration while going through the rest of this exercise, agreed

an older sysadmin practice of using tools like logrotate would generally handle this in a similar way, usually just by `(max files x max file size)` as a calculation.

**Comment by HirazawaUi** [comment]
> I don't think this would be super beneficial for "external log shipper" purposes, unless you're talking a substantially large size of a log in a short duration (theoretically possible). from a log shippers perspective, that file becoming compressed is another opportunity for that shipper to need to rescan for available log files. being most log shippers tend to do this rescan between 1-60 second intervals, the benefits may not be there...

Introducing a delayed cleanup mechanism is definitely an effective way to log reliability, and I think your suggestion will be very useful. However, it also comes with some challenges:

1. Disk space filling up during the delay period
2. Task loss due to Kubelet restarts

This might introduce some complexity, but I’ll include it in the proposal and bring it to the sig-node meeting in the future to gather more feedback.

**Comment by harshanarayana** [comment]
Cross posting my earlier post here with some more context. https://github.com/kubernetes/kubernetes/issues/123279#issuecomment-1951759359

This is mainly in context of `Rotation and Compression of Logs from Deleted Pods: `

We  have been running into issues with Debugging failures on the short lives Job based pods for a bit and since we manage our own version of k8s, we had to implement something similar to this. Here is what we did. 

1. We added a feature flag that can toggle the retention of terminated pod logs in the Kubelet that can be toggled on or off.
2. We added a configuration path to a directory where the terminated pod logs can be relocated to. If Not provided, this would default to `/var/log/terminated-pod-logs/<service-name>`
3. The hierarchy of the logs in this dir is the same as that of the `/var/log/pods`. This would ease the debugging pain. 


In order to simplify the complications of the disk space problems caused by the retention of the terminated logs, we did a few things. 

1. The actual log retention after being enabled in kubelet is done per service basis. With the helm of a configuration. That configuration being `kubernetes.io/terminated-retention-name`. Once those pods that had this annotation would reap the benefits of this feature. 
2. We also added an optional annotation `kubernetes.io/terminated-retention-count` that can be used to configure how many such terminated pod logs are to be retained per service basis. This way kubelet would auto GC them while relocating. 

We also added an external log rotate based cleanup that can routinely monitor the usage of this terminated pod log dir and wipe it clean when it exceeded a certain threshold.

In order to make sure we managed to ensure the retention of these during abrupt restart of kubelet during the cleanup workflow, we had to add a few hacked-up logic to the `/var/log/pods/...` dir to store some additional metadata. So that even if the kubelet were to restart and it had to just wipe the log dir., it would still do it's best to retain the terminated logs. 


```bash
$ find /var/log/pods/terminated-pod-logs/
/var/log/pods/terminated-pod-logs/
/var/log/pods/terminated-pod-logs/dummysvc
/var/log/pods/terminated-pod-logs/dummysvc/.metadata
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-gt2xm_58b60bbd-c7a5-46d8-836c-d1e43699511a
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-gt2xm_58b60bbd-c7a5-46d8-836c-d1e43699511a/dummysvc
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-gt2xm_58b60bbd-c7a5-46d8-836c-d1e43699511a/dummysvc/0.log
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-jxtlk_b746ba2f-6582-45a0-9065-de7473682223
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-jxtlk_b746ba2f-6582-45a0-9065-de7473682223/dummysvc
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-jxtlk_b746ba2f-6582-45a0-9065-de7473682223/dummysvc/0.log
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-phv4l_16fcdc2b-6a0f-40d2-8ae9-5998a4b82dcf
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-phv4l_16fcdc2b-6a0f-40d2-8ae9-5998a4b82dcf/dummysvc
/var/log/pods/terminated-pod-logs/dummysvc/tardis_dummysvc-phv4l_16fcdc2b-6a0f-40d2-8ae9-5998a4b82dcf/dummysvc/0.log

**Comment by HirazawaUi** [comment]
> In order to simplify the complications of the disk space problems caused by the retention of the terminated logs, we did a few things.
> 
> 1. The actual log retention after being enabled in kubelet is done per service basis. With the helm of a configuration. That configuration being `kubernetes.io/terminated-retention-name`. Once those pods that had this annotation would reap the benefits of this feature.
> 2. We also added an optional annotation `kubernetes.io/terminated-retention-count` that can be used to configure how many such terminated pod logs are to be retained per service basis. This way kubelet would auto GC them while relocating.

This undoubtedly introduces a degree of flexibility, enabling better alignment with the needs of different services, avoiding a rigid "one-size-fits-all" global configuration, and enhancing the system's adaptability.

However, it also introduces some complexity no offense intended such as the following theoretical considerations:

  1. Kubelet would need to parse and manage distinct configurations for each service, which could increase the difficulty of development and maintenance. If misconfigured, this might result in unintended log loss or resource wastage.
  2. With some services having long log retention periods and others short ones, this could complicate troubleshooting across services.

I’m not certain whether you would encounter these issues, though.


It seems you aren’t using an "log collection tools"? such as Filebeat or similar tools, Or perhaps you are, but I haven’t fully understood it correctly. If you aren’t using an "log collection tools," why haven’t you considered making log compression an optional feature?

Lastly, thank you so much for sharing your practical experience—it’s genuinely helpful!

**Comment by mmerickel** [comment]
This is naive - I'm still learning who is responsible for what between containerd and kubelet. It seems that there's a hardcoded protocol between them that the log files will be paths on disk. 
I understand that containerd+kubelet want the log files as their protocol so that kubelet can read the logs and serve them via the kubernetes log endpoints. 

1) Is there a reason that kubelet takes responsibility for cleaning up logs versus traditional methods for log file handling like deferring to logrotate?

We use vector for sourcing dmesg/containerd/kubelet logs (journald) and pod logs (raw files on /var/log/pods). It'd be cool if for starters containerd could optionally just dump the pod logs into journald to completely get itself out of the business of log file management. Then things could be written to disk too but we'd just ignore them and kubelet could delete them aggressively as it does today with no changes there.

2) Maybe containerd could just have an option to tee the logs to journald as well as to the current files, such that systems could enable that if they want and then pull the logs directly from journald just like most other logs in the system. This would get out of the business of caring about inotify and file-based problems related to external systems that want to watch/ingest the logs.

**Comment by HirazawaUi** [comment]
> I'm still learning who is responsible for what between containerd and kubelet. It seems that there's a hardcoded protocol between them that the log files will be paths on disk.

The interaction between containerd and kubelet is based on CRI, and the log file path is passed through CRI as part of the protocol, rather than being hard-coded.


> 1. Is there a reason that kubelet takes responsibility for cleaning up logs versus traditional methods for log file handling like deferring to logrotate?

There’s a reason why kubelet is responsible for log cleanup. In Kubernetes, Pods are dynamic, frequently created, deleted, or migrated, so the lifecycle of log files must be tightly coupled with the Pods. As the direct manager of Pods, kubelet can precisely clean up logs based on Pod status and configuration. On the other hand, logrotate is a general-purpose tool that struggles to adapt to such dynamic demands. Forcing its use might lead to log management chaos, or perhaps my understanding of logrotate isn’t entirely accurate?


> 2\. Maybe containerd could just have an option to tee the logs to journald as well as to the current files, such that systems could enable that if they want and then pull the logs directly from journald just like most other logs in the system. This would get out of the business of caring about inotify and file-based problems related to external systems that want to watch/ingest the logs.

Container logs are based on the file, supporting tools like Filebeat for log collection. Redirecting them to journald would limit this flexibility. In many private cloud environments, logs from kubelet and containerd are redirected to log files to avoid confusion, rather than being output to journald. Sending container logs to journald would undoubtedly make everything more complicated.


In summary, it seems you want to manage container logs using the log management approach of traditional services, which may not align well with the current ecosystem or the goals of this issue.

**Comment by mmerickel** [comment]
The goal is to guarantee ingestion of the pod logs to an external system. Tools like journald provide this whereas currently we are relying on disparate systems racing against each other with no knowledge of the other to prevent the races everyone is experiencing. Even if we slow down kubelet to add a delay between when a pod ends and when it deletes the logs it doesn’t guarantee the external system like filebeat/vector/fluentd had time to ingest the logs from the file on disk before they are gone. It certainly helps but still leaves us on the outside scrambling / polling / inotifying / listening to find the files in time and to read them fully in time. This buffering / muxing of log sources to sinks it what things like journald/ syslog / sockets are great at. I’m not saying it has to be done that way but it’s an example of something that would work and manage the lifecycle more cleanly without the current race conditions.

**Comment by HirazawaUi** [comment]
> it what things like journald/ syslog / sockets are great at. I’m not saying it has to be done that way but it’s an example of something that would work and manage the lifecycle more cleanly without the current race conditions.

Thank you for your explanation, I think it makes a lot of sense.
After a quick search, it seems that crio already supports what you mentioned about outputting container logs to journald.
ref:  https://github.com/cri-o/cri-o/pull/2086

However, I haven’t explored this feature yet. If you’re interested, perhaps you could try switching from containerd to crio to test whether it meets your needs.

Beyond that, I’ve also come across some additional features that might be useful. I’ll continue organizing them a bit later.

**Comment by kannon92** [comment]
This reminds me of this ticket: https://github.com/kubernetes/kubernetes/issues/129975

One can always disable log rotation (I believe) and store the logs. Kubelet also has configurable location for logs.

**Comment by CpuID** [comment]
re the journald suggestion, one consideration is that Kubernetes runs on both Windows and Linux these days, which would fragment things even further (being journald tends to be not a Windows component). maintaining a file-based solution likely makes it easier to support both OSes successfully

> The goal is to guarantee ingestion of the pod logs to an external system.

I've been thinking a bit about the problem more, especially around filebeat's autodiscover mechanism and the fact that's a K8s-specific solution, but doesn't solve for other container schedulers such as ECS, etc.

It's made me wonder wether relying on inotify or some kernel-level construct (plus a Windows equivalent?) is the right answer, to be scheduler agnostic? Then the log shipping tools all just need to standardise on inotify style support for "watching for new files"?

**Comment by harshanarayana** [comment]
> Kubelet would need to parse and manage distinct configurations for each service, which could increase the difficulty of development and maintenance. If misconfigured, this might result in unintended log loss or resource wastage.

There is no fancy parsing needs. It is just two additional annotations used only during the rotation phase to extract the service names to be used for deciding the dir structure. 

> With some services having long log retention periods and others short ones, this could complicate troubleshooting across services.

Yes and no. The needs for individual services are / can be different. Not all services need the same amount of longs to be retained. But I do agree it makes easy to manage if that was done by default. We default to 3 terminated pod per service but allow one to configure with an override if they want to do so. 

> such as Filebeat or similar tools, Or perhaps you are, but I haven’t fully understood it correctly. If you aren’t using an "log collection tools," why haven’t you considered making log compression an optional feature?

We do use fluent-bit and ES for storage. However, there are short lives pods like the one from Jobs that have shorter TTL and can get deleted before they are scraped at times. Also, considering we run on On-Prem, the retention on ES is limited. So, having the ability to store the logs in host enables us to store them in external storage location such as that of NFS mounted path. 

@HirazawaUi 


> Maybe containerd could just have an option to tee the logs to journald as well as to the current files, such that systems could enable that if they want and then pull the logs directly from journald just like most other logs in the system.

[This is one thing I terribly miss from Docker days](https://docs.docker.com/engine/logging/drivers/journald/). Where one could easily write a log driver or different mechanism to manage logs. But from whatever I could find, containerd has no such plugin support for log management. https://github.com/containerd/containerd/issues/4798

But as we speak I am working on an internal fork that can provide this mechanism where the logs are/can be routed the journald instead of file so that we can manage the logs there and rotate using a universal log rotate configuration at the Journal level. This also enables better integration with things like log drains on rsyslog. 


> Even if we slow down kubelet to add a delay between when a pod ends and when it deletes the logs it doesn’t guarantee the external system like filebeat/vector/fluentd had time to ingest the logs from the file on disk before they are gone.

This is the reason we decided to move the logs to a different path first and then rotate from there with a custom configuration outside of kubelet. That guarantees us we can get the logs where/when we want with better certainty. 

@mmerickel 

> One can always disable log rotation (I believe) and store the logs. Kubelet also has configurable location for logs.

But there is no way to prevent it from GC'ing the logs of a pod being deleted. Think of a bunch of jobs doing some replica adjustment on the DB/setting up some configuration DB on need based workflows, the jobs come and go and when the pods are deleted, so are the logs from those pods. Unless one configures a really high TTL, the pods can get GC'ed before the logs from them are scraped by tools like file-beat or such. Right ?

@kannon92

**Comment by harshanarayana** [comment]
I am definitely open to a KEP. I can start a KEP with the current Implementation details of what we did and we can discuss how sensible / logic some of those are and see where we can go from there.

**Comment by kannon92** [comment]
My suggestion would be to try and bring up your usecase and maybe a sketch of the design to sig-node before going directly into the KEP.

It seems there is a need for configurations of log management and we should aim to address that.

**Comment by harshanarayana** [comment]
That works for me @kannon92

**Comment by HirazawaUi** [comment]
Ah, I'm very glad you're willing to share your thoughts. Today, I found some fragments of our past discussions about it in the historical issues and PRs (it’s more complicated than I expected). Perhaps you've already seen some of them, but if you haven’t, maybe you could use their content as a reference to implement your proposal.

https://github.com/kubernetes/kubernetes/pull/33111 : Proposes a new logging mechanism for kubelet-cri, including a "logs" object and logging policies


https://github.com/kubernetes/kubernetes/issues/24677 : Discuss integrating Kubernetes logs with Splunk, including node and application logs, and seek solution suggestions.


https://github.com/kubernetes/kubernetes/issues/13846 : On the future of log processing in Kubernetes, discussing current issues with node-level log management and proposing alternatives to using a cluster-level logging infrastructure.


https://github.com/kubernetes/kubernetes/pull/13010 : Pull request to add LogDir functionality for managing pod log files for easy log collection and management and integration with Fluentd

https://github.com/kubernetes/kubernetes/issues/17183 : Discuss the log management roadmap for Kubernetes, including tools such as kubectl logs, fluentd, ELK, and the Docker log plugin.

**Comment by harshanarayana** [comment]
https://github.com/harshanarayana/containerd/tree/enable-conffigurable-sys-log

@kannon92  @HirazawaUi 

I am working on getting a Doc Created with some of the details and working on some PoC on what can be done. Above is the first version of that. This enables a way for containerd to write logs to journald along with its usual write to file workflow. This way, the logs from terminated pods will still be present in Journal. This is only one crude way to achieve things. Someone already brought up a mechanism similar to this and this is an attempt to show that it is possible. But this will be highly linked to Linux + systemd being present on the system for it to function. 

Other one I am working on is a PoC with a pluggable logger via the fifo/pipes that can work in Windows/*nix all the same. Will share a PoC of the same in a day or two. Please take a look at the above one and see if this is something to be included in the discussions.

**Comment by harshanarayana** [comment]
https://github.com/harshanarayana/containerd/commits/enable-log-plugins/

As mentioned above, this is a crude implementation of the containerd plugin for logging via the gRPC. Took this route instead of the pipes/fifo etc so that this can work across the board in both windows and unix the same way. 

> This is still to be tested posting here for reference for others to provide feedback,

**Comment by harshanarayana** [comment]
That's the next one I have in my plan.. Move store and rotation to containerd entirely.. Then update the cri to have a new api that kubelet can use for getting logs for a given pod/container combo. Then containerd can have any log plugin that may ormay not even write log to host directory.. But provide logs for kubecrl logs via the cri api.. Would that be too much of indirecrion?

**Comment by agowa** [comment]
/reopen 

We've basically the same need and it proves an issue over and over again in daily operation for debugging problems with esp. intermittently failing pods.

From the concerns stated in the initial post one suggestion how about the possibility to add a custom tag to a pod that only when present and set to true would cause the kubelet to skip over deleting it when it enters a failed state?

Alternatively to a bool it could be a time interval in days. Aka the failed pod is kept around as is for up to the specified amount of days but without blocking the self-healing/recreation logic. This however may make another pod status like "pending-cleanup" necessary as otherwise it could run into issues with persistent volume claims (not sure if they'd be seen as available again when they're attached to a pod in failed state that hasn't been deleted yet. And we can't just remove the reference either, as that would change the state the pod was in when it failed and mislead ops while investigating)
